proposals/NNNN-swiftpm-additional-plugin-apis.md

```diff
@@ -0,0 +1,450 @@
+    /// they are declared in the package manifest. The product will contain the
+    /// transitive closure of the these targets and their depdendencies.
+    public let targets: [Target]
+}
+
+/// Represents an executable product defined in a package.
+public class ExecutableProduct: Product {
+    /// The target that contains the main entry point of the executable. Every
+    /// executable product has exactly one main executable target. This target
+    /// will always be one of the targets in the product's `targets` array.
+    public let mainTarget: Target
+}
+
+/// Represents a library product defined in a package.
+public class LibraryProduct: Product {
+    /// Whether the library is static, dynamic, or automatically determined.
+    public let type: LibraryType
+}
+
+/// A type of library product.
+public enum LibraryType {
+    /// Static library.
+    case `static`
+
+    /// Dynamic library.
+    case `dynamic`
+
+    /// The type of library is unspecified and will be determined at build time.
+    case `automatic`
+}
+
+/// Represents a single target defined in a package. Specializations represent
+/// different types of targets.
+public class Target {
+    /// The name of the target, as defined in the package manifest. It's unique
+    /// among the targets of the package.
+    public var name: String
+    
+    /// The absolute path of the target directory in the local file system.
+    public var directory: Path
+    
+    /// Any other targets on which this target depends, in the same order as
+    /// they are specified in the package manifest. Conditional dependencies
+    /// that do not apply have already been filtered out.
+    public var dependencies: [Dependency]
+
+    /// Represents a dependency of a target on a product or on another target.
+    public enum Dependency {
+        /// A dependency on a target in the same package.
+        case target(Target)
+
+        /// A dependency on a product in another package.
+        case product(Product)
+    }
+}
+
+/// Represents a target consisting of a source code module, containing either
+/// Swift or source files in one of the C-based languages.
+public class SourceModuleTarget: Target {
+    /// The name of the module produced by the target (derived from the target
+    /// name, though future SwiftPM versions may allow this to be customized).
+    public var moduleName: String
+
+    /// The directory containing public C headers, if applicable. This will
+    /// only be set for targets that have Clang sources, and only if there is
+    /// a public headers directory.
+    public var publicHeadersDirectory: Path?
+
+    /// The source files that are associated with this target (any files that
+    /// have been excluded in the manifest have already been filtered out).
+    public var sourceFiles: FileList
+}
+
+/// Represents a target describing a library that is distributed as a binary.
+public class BinaryLibraryTarget: Target {
```

Comment:
Could potentially be interesting to retain the original source here (local vs. remote and if remote, what was the origin URL)? For example, I could imagine a plug-in that verifies that only allowed origins are used for binaries in the package graph.

Reply:
Yes, that’s a good point.  Maybe we just should expose all of it, e.g. SHA etc, for logging purposes for example.  I could see code generation plugins also wanting to produce some kind of manifest of what was included, for tracking purposes.

Reply:
Added this.  See what you think.